File: lclang/src/main/java/lclang/libs/std/classes/OutputClass.java
package lclang.libs.std.classes;

import lclang.exceptions.LCLangIOException;
import lclang.libs.lang.classes.ArrayClass;
import lclang.libs.lang.classes.LCClass;
import lclang.libs.lang.classes.LibraryClass;
import lclang.libs.lang.classes.VoidClass;
import lclang.libs.lang.classes.numbers.IntClass;
import lclang.libs.lang.classes.numbers.NumberClass;
import lclang.methods.Method;
import lclang.types.CallableType;
import lclang.types.Type;
import lclang.types.Types;

import java.io.IOException;
import java.io.OutputStream;
import java.io.PrintStream;
import java.io.UnsupportedEncodingException;
import java.util.Collections;
import java.util.List;

public class OutputClass extends LibraryClass {
    public static final String name = "Output";
    public static final OutputClass instance = new OutputClass();
    public static final Type type = instance.classType;
    public PrintStream printer;

    private OutputClass() {
        super(name);

        constructor = method((caller, args) -> {
            Method method = (Method) args.get(0);
            return new OutputClass(new OutputStream() {
                @Override
                public void write(int b) {
                    method.call(caller, Collections.singletonList(IntClass.get(b)));
                }
            });
        }, CallableType.get(IntClass.TYPE, VoidClass.type), type);
    }

    public OutputClass(OutputStream stream) {
        this();
        try {
            printer = new PrintStream(stream, false, "UTF-8");
        } catch (UnsupportedEncodingException ignored) {}

        globals.put("flush", voidMethod((caller, args) -> printer.flush()));
        globals.put("close", voidMethod((caller, args) -> printer.close()));
        globals.put("accept", voidMethod((caller, args) -> {
            InputClass input = (InputClass) args.get(0);

            try {
                String line;
                while ((line = input.reader.readLine())!=null)
                    printer.println(line);
            }catch (IOException e) {
                throw new LCLangIOException(e.getMessage(), caller);
            }

        }, InputClass.type));

        globals.put("write", voidMethod((caller, args) -> printer.write(args.get(0)
                .cast(NumberClass.class)
                        .number.byteValue()), NumberClass.TYPE));
        globals.put("print", voidMethod((caller, args) ->
                printer.print(args.get(0).toString(caller)), Types.ANY));
        globals.put("println", voidMethod((caller, args) ->
                printer.println(args.get(0).toString(caller)), Types.ANY));
        globals.put("printf", voidMethod((caller, args) -> {
            String main = args.get(0).toString(caller);
            List<LCClass> classes = ((ArrayClass) args.get(1)).value;
            String[] printArgs = new String[classes.size()];
            for (int i = 0; i < classes.size(); i++) {
                printArgs[i] = classes.get(i).toString(caller);
            }

            printer.printf(main, (Object[]) printArgs);
        }, Types.ANY, ArrayClass.type));
    }
}

File: lclang/src/main/kotlin/lclang/libs/std/classes/OutputClass.kt
package lclang.libs.std.classes

import lclang.LCClass
import lclang.constructor
import lclang.libs.lang.ArrayClass
import lclang.libs.lang.IntClass
import lclang.methods.Method
import lclang.types.CallableType
import lclang.types.Types
import lclang.void
import java.io.OutputStream
import java.io.PrintStream

const val OUTPUT_CLASSNAME = "Output"
class OutputClass(): LCClass(OUTPUT_CLASSNAME) {
    override var constructor: Method? = constructor(listOf(CallableType(arrayOf(Types.INT)))) {
        val method = it[0] as Method
        OutputClass(object : OutputStream() {
            override fun write(b: Int) {
                method.call(this@constructor, listOf(IntClass(b).asValue()))
            }
        })
    }

    constructor(output: OutputStream) : this() {
        val printer = PrintStream(output)

        globals["println"] = void(Types.ANY) { printer.println(it[0]) }
        globals["print"] = void(Types.ANY) { printer.print(it[0]) }
        globals["printf"] = void(Types.ANY, Types.ARRAY) { list -> printer.printf(list[0].toString(),
            * (list[1] as ArrayClass).list.map { it.get(this) }.toTypedArray())}
        globals["flush"] = void { printer.flush() }
        globals["close"] = void { printer.close() }
        globals["accept"] = void(Types.MagicType(INPUT_CLASSNAME)) {
            val input = it[0] as InputClass
            while (input.scanner.hasNextLine())
                printer.println(input.scanner.nextLine())
        }
    }
}